object_db.go

```diff
@@ -256,52 +269,33 @@ func (d *ObjectDatabase) encodeBuffer(object Object, buf io.ReadWriter) (sha []b
 // save writes the given buffer to the location given by the storer "o.s" as
 // identified by the sha []byte.
 func (o *ObjectDatabase) save(sha []byte, buf io.Reader) ([]byte, int64, error) {
-	n, err := o.s.Store(sha, buf)
+	n, err := o.rw.Store(sha, buf)
 
 	return sha, n, err
 }
 
 // open gives an `*ObjectReader` for the given loose object keyed by the given
 // "sha" []byte, or an error.
 func (o *ObjectDatabase) open(sha []byte) (*ObjectReader, error) {
-	f, err := o.s.Open(sha)
-	if err != nil {
-		if !os.IsNotExist(err) {
-			// If there was some other issue beyond not being able
-			// to find the object, return that immediately and don't
-			// try and fallback to the *git.ObjectScanner.
-			return nil, err
-		}
-
-		// Otherwise, if the file simply couldn't be found, attempt to
-		// load its contents from the *git.ObjectScanner by leveraging
-		// `git-cat-file --batch`.
-		if atomic.LoadUint32(&o.closed) == 1 {
-			return nil, fmt.Errorf("gitobj: cannot use closed *pack.Set")
-		}
+	if atomic.LoadUint32(&o.closed) == 1 {
+		return nil, fmt.Errorf("gitobj: cannot use closed *pack.Set")
+	}
 
-		packed, err := o.packs.Object(sha)
+	for _, s := range o.ro {
+		f, err := s.Open(sha)
 		if err != nil {
+			if errors.IsNoSuchObject(err) {
+				continue
+			}
 			return nil, err
 		}
-
-		unpacked, err := packed.Unpack()
-		if err != nil {
-			return nil, err
+		if s.IsCompressed() {
+			return NewObjectReadCloser(f)
+		} else {
```

Comment:
Nit: some linters will complain that an `else` block that follows an `if` with an unconditional return is redundant, and should be rewritten as:

```go
if foo {
        return bar
}
return baz
```

Reply:
Yeah, thanks for pointing this out.  I always do this in C, but for some reason I haven't yet convinced myself to do it in Go.